staging/src/k8s.io/api/networking/v1/types.go

```diff
@@ -201,3 +201,172 @@ type NetworkPolicyList struct {
 	// Items is a list of schema objects.
 	Items []NetworkPolicy `json:"items" protobuf:"bytes,2,rep,name=items"`
 }
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// Ingress is a collection of rules that allow inbound connections to reach the
+// endpoints defined by a backend. An Ingress can be configured to give services
+// externally-reachable urls, load balance traffic, terminate SSL, offer name
+// based virtual hosting etc.
+type Ingress struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// Spec is the desired state of the Ingress.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	// +optional
+	Spec IngressSpec `json:"spec,omitempty" protobuf:"bytes,2,opt,name=spec"`
+
+	// Status is the current state of the Ingress.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	// +optional
+	Status IngressStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// IngressList is a collection of Ingress.
+type IngressList struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// Items is the list of Ingress.
+	Items []Ingress `json:"items" protobuf:"bytes,2,rep,name=items"`
+}
+
+// IngressSpec describes the Ingress the user wishes to exist.
+type IngressSpec struct {
+	// The default backend capable of servicing requests that don't match any
+	// rule. At least one of 'defaultBackend' or 'rules' must be specified. This field
+	// is optional to allow the loadbalancer controller or defaulting logic to
+	// specify a global default.
```

Comment:
Should we more strongly encourage this to be populated?  We should at least say something like "If this is not specified, any request which does not match a rule will be handled in an implementation-defined way (e.g. a generic 404 or a 5xx)."

Reply:
Not sure since the last comment in here suggests that there was an original purpose for it to be optional. I added your suggested comment.

Reply:
I think we can't really make it required because of round-trip, but in practice the global default has turned out to be a DISASTER.  Let's NOT make this mistake again.

@rramkumar1 @aledbf  One thing we might do is throw an event from ingress controllers warning that no defaultBackend is set, and that unhandled URLs may fall through to the global default.

Reply:
We dont have a mechanism to feed "warnings" back through API ops (sadly)

Reply:
> We dont have a mechanism to feed "warnings" back through API ops (sadly)

nowhere close to available yet, but I started looking into doing that via warning headers in https://github.com/kubernetes/kubernetes/pull/73032